### PR TITLE
Apply improvements to the kind polymorphism design

### PR DESCRIPTION
**Pull request by goldfirere**
This PR encompasses little improvements suggested by various colleagues (as noted in commit messages). I recommend reviewing commit by commit.

Review requested from @stedolan, who gave some very helpful feedback and has already looked at these rules in detail.

### CODE REVIEW

**Comment by stedolan** [comment]
I haven't read it all yet but the new `SUB` and `MB_*` rules look good to me. It's missing one almost trivial rule, though: I need to be able to compute joins of concrete modes on the right, by something like:
```
Γ ⊢ mbi ≤ modal_bound₂_Ξ ⊔ (mode (m₁_Ξ ⊔ m₂_Ξ))
------------------------------------------------ MB_JOIN_R
Γ ⊢ mbi ≤ modal_bound₂_Ξ ⊔ (mode m₁_Ξ) ⊔ (mode m₂_Ξ) 
```
This rule is actually doing something (barely): you might have mbi less than the join of two concrete modes appearing on the right, but not less than either by itself. So you need to actually compute the join on the right in order to compare it to the left with MB_MODE.

**Comment by goldfirere** [comment]
I think I've addressed the concern above, but please double-check, as the rule I wrote is not exactly the rule you suggested.

Would you mind taking a deeper look at the rest of the changes here to see if we can merge this, @stedolan? Thanks!

File: ocaml/jane/doc/proposals/kind-inference.md [hunk 1]
```diff
@@ -197,22 +200,23 @@ k = κ ∈ Γ
...
-Γ ⊢ kind_of_ τ ↠ layout_of τ; ⟪⊥_Ξ with τ⟫
-
 Γ ⊢ jkind ↠ κ
 ------------------------------------------------- K_MOD
 Γ ⊢ jkind mod modes ↠ lay(κ); ⟪Ξ(κ) ⊓ ⨅ extract(Ξ, modes)⟫
-  (* CR reisenberg: This rule isn't right if κ already has [with] constraints; probably
-     should just forbid that syntactically. Not right: we can't meet out
-     the with-added modes because we don't really know what they are.
-     So we'd have to retain the [mod] clause just like we retain the
-     [with] clause, which is terrible. *)
```
Comment (stedolan) [inline ocaml/jane/doc/proposals/kind-inference.md:211]: I have the same question, and I don't understand how it was resolved by this change. What does happen if κ already has [with] constraints?

**Comment by stedolan** [comment]
(apologies this took so long to get to)

The new MB_* rules look good to me. I'm curious about what's going on with the deleted CR I commented on above, though: essentially, you have a join operation (`with`) and a meet operation (`mod`), and I would like to understand whether and how they interact.

File: ocaml/jane/doc/proposals/kind-inference.md (hunk 1 above)
Comment (goldfirere) [reply ocaml/jane/doc/proposals/kind-inference.md:211]: The user-written syntax indeed prevents `with` constraints from ending up in the \kappa here. But the system does not track this. I have split out \chi from \kappa, where \chi does not have `with` constraints. This is a nice improvement. I also may have noticed a few other small bugs that I fixed while taking a pass through.

**Comment by lpw25** [comment]
Maybe this is discussed in the doc already, but perhaps it is worth me pointing out that `mod` is not really a meet. It represents a morphism on modes that is either meet-with-constant for comonadic modes or join-with-constant for monadic modes.

**Comment by goldfirere** [comment]
The doc is silent about mode-crossing (somewhat paradoxically, because it describes a system that's all about mode-crossing). It just defines subkind operations. My last commit reverses the order of elements in the monadic axes and now includes some commentary about this potentially-confusing point.

**Comment by stedolan** [comment]
> It's not a general binary meet, but meet-with-constant is still a meet. In particular, the syntax `k mod observing with σ` seems to be allowed in a jkind_scheme, and I don't fully understand what it means or how it is converted to a modal_bound. (Does the join distribute over the meet?)

Ah, from a closer look at `K_MOD` I see that (a) `mod` is completely evaluated during initial translation to `χ` and (b) the left hand side of `mod` must also be a constant (a `const_layout`, another `mod` expression, or a kind abbrevation `k`, but _not_ `kind_of_`). This certainly simplifies things, but will this simplicity survive the addition of mode/kind polymorphism? (I have been assuming that eventually kind abbreviations `k` will generalise to include kind variables that cannot be trivially expanded).

**Comment by goldfirere** [comment]
I think @lpw25's comment meant to say that `mod` is not *always* a meet. That is, the interpretation of `mod` with monadic axes is not a meet at all, but a `subtract`.

Mode polymorphism does not intersect the design in this doc. Mode polymorphism is a feature of terms, not of types. Maybe someday we'll allow types to be parameterized on modes -- and then perhaps we'll allow the modes to appear in kind-level modal bounds -- but there are no plans to add this, and I don't think we need to make the design robust against that possibility now.

There are additionally no plans for kind variables of the form you're worried about. Perhaps unhelpfully, the ideas in the doc as written are called "kind polymorphism". I clearly need a different name for this suite of features, because this is a common source of confusion. Suggestions welcome.

**Comment by goldfirere** [comment]
Are there remaining open questions here? Or should we merge?

**Review by stedolan — APPROVED** [review]

**Pull request merged**